### ISSUE TITLE
Unable to Concurrently Backup to Multiple Repositories

### ISSUE THREAD

**Issue by andrewlecuyer** [pgbackrest/pgbackrest#1406]
1. pgBackRest version: **v2.33**

2. PostgreSQL version: **v13.2**

3. Operating system/version - if you have more than one server (for example, a database server, a repository host server, one or more standbys), please specify each:
A single database host and a single repository host, each running inside of it's own container built on top of CentOS v8.3.

4. Did you install pgBackRest from source or from a package?
Package.

5. Please attach the following as applicable:

`pgbackrest.conf`:

```ini
[global]
log-path=/tmp
repo1-path=/pgbackrest/repo1
repo2-path=/pgbackrest/repo2

[db]
pg1-host=test-instance1-j7kn-0.test-pods
pg1-path=/pgdata/pg12
pg1-port=5432
pg1-socket-path=/tmp
```

`postgresql.conf`:

```ini
archive_command = 'pgbackrest --stanza=db archive-push %p'
archive_mode = 'on'
listen_addresses = '*'
max_wal_senders = '10'
port = '5432'
wal_level = 'logical'
```

Errors (as seen when attempting to create a second backup after another backup is already in progress):

```bash
$ pgbackrest backup --stanza=db --repo=2
ERROR: [050]: unable to acquire lock on file '/tmp/pgbackrest/db-backup.lock': Resource temporarily unavailable
       HINT: is another pgBackRest process running?
```

6. Describe the issue:

With multiple pgBackRest repositories defined, it is not appear to be possible to take a backup for multiple repositories at the same time.  Specifically, attempting to initiate a second backup for a second repo results in the `Resource temporarily unavailable` error shown above.

Therefore, is it possible to configure pgBackRest to allow concurrent backups for different repositories?  Or do I need to ensure the various backups across the various repositories do not overlap?  And if so, are there any general suggestions for scheduling backups across multiple repositories to avoid potential conflicts with locks?

The desire for concurrent backups stems from the potential complexity associated with scheduling backups across multiple repos (now up to four) that don't overlap - especially when available windows for performing backups are small, and other factors (e.g. the size of the database, unexpected failures and/or latency during backups, etc.) can increase the likelihood of overlapping backup attempts, and therefore failed backups for certain repos.

Thanks!

**Comment by pgstef** [comment]
Hi,

**2.33** is the first release introducing this feature. And at this point, I don't think it's possible to take 2 backups of the same stanza at the same time. And IMHO, that's a good thing. You probably don't want to read all your production data twice at the same time.

What could be super interesting would be to copy backup sets from one repository to another, not having to read the same production data again.

For now, probably a simple cron like `pgbackrest backup --stanza=db --repo=1 ; pgbackrest backup --stanza=db --repo=2` would allow to take the backup to repo2 just after the one in repo1.

Kind Regards

**Comment by pgstef** [comment]
To be complete, the repo-to-repo backup feature is in the project [backlog](https://github.com/pgbackrest/pgbackrest/projects/2#card-59620607).

Kind Regards

**Comment by andrewlecuyer** [comment]
@pgstef thank you for the response, and the suggestion for scheduling back-to-back cron jobs.  I will keep an eye on the feature you referenced, as it does sounds like a effective way to address my specific use case (albeit in a bit of a different way).

Should I assume repo-to-repo will be the preferred solution for addressing this use case within pgBackRest moving forward (vs. the ability to concurrently backup)?  Additionally, is the plan to allow multiple repositories to perform a repo-to-repo backup from the same repository concurrenty?

Thanks again.

**Comment by pgstef** [comment]
@dwsteele might be in a better place to answer that question but I can imagine we'd need to review the locking system to allow concurrency there. And I'm not sure we'd want to do that.

**Comment by dwsteele** [comment]
My feeling is we'd need to allow concurrent backups both from a PG host and from a source repo. A destination repo would only allow one backup at a time, of course. But we won't really know until we get into it and see what is possible.

**Comment by turnstep** [comment]
I'm guessing that the solution `pgbackrest backup --stanza=db --repo=1 ; pgbackrest backup --stanza=db --repo=2` is only for cases in which one of the repos is "local" and one is "S3" etc.? Because that won't work on two separate VM repos, but I suppose some sort of home-rolled semaphore lock would. Anyway, big +1 to the idea of repo2 being able to pull the files directly from repo1 (or as many as possible, similar to `backup-standby`).

**Comment by dwsteele** [comment]
> Because that won't work on two separate VM repos

Not sure what you mean by this?

**Comment by rubik** [comment]
Is there any update on this issue? We have a similar configuration: a local repository and a GCS bucket repository. Currently, it seems the only way to backup to both is to run the `backup` command twice, one for each repository. However, that's very wasteful and it takes a very long time as it's copying the database data twice in a row. The backups are going to be very similar too.

A better solution for this particular case would be to copy the backup from the local repository directly to GCS. As I understand, that's what the repo-to-repo feature is about. As it's still in the backlog, are there any alternatives?

One that comes to mind is to remove the remote repository from pgBackRest's configuration and use some Cron jobs to push local backups to the GCS bucket. However, pgBackRest won't know about the remote repository anymore. In case of a recovery, we'd need to edit the configuration to add the remote repository, but I imagine there could be problems with the fact that pgBackRest didn't create the repository directly and there might be metadata mismatches.

**Comment by dwsteele** [comment]
> Is there any update on this issue?

It is still on our radar but I can't give you any kind of ETC. It requires multiple changes in locking, which are tricky, and I haven't had the time to focus on it.

> One that comes to mind is to remove the remote repository from pgBackRest's configuration and use some Cron jobs to push local backups to the GCS bucket. However, pgBackRest won't know about the remote repository anymore.

Well, you could configure that remote repo for some purposes, as long as archive-push doesn't see the config. So, for instance:
```
[stanza:restore]
repo2-path=...

[stanza:archive-get]
repo2-path=...
```
I haven't tried this but looks like it would work OK.

> but I imagine there could be problems with the fact that pgBackRest didn't create the repository directly and there might be metadata mismatches.

It should be fine as long as you make an exact copy of the local repo to the remote. If you try to cherry-pick parts then yeah, there will be a problem.

**Comment by dwsteele** [comment]
This has been implemented in #2389 so closing this issue.
